### PR TITLE
[autograd] add tests for simple reentrant and stackoverflow escape

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33157 [autograd] enable graph level thread parallelism on CPU
* **#35259 [autograd] add tests for simple reentrant and stackoverflow escape**

This PR added tests as part of https://github.com/pytorch/pytorch/issues/34367

It covers:
Re-entrant -> Test simple re-entrant
Re-entrant -> Test stack overflow escape mechanism

Differential Revision: [D20611828](https://our.internmc.facebook.com/intern/diff/D20611828)

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -5721,6 +5776,49 @@ def test_copy_(self, device):
...
+
+            @staticmethod
+            def backward(ctx, grad_output):
+                with torch.enable_grad():
+                    new_param = ctx.output_var.detach().requires_grad_()
+                    if ReentrantFunc._cpu_mode:
+                        new_param = new_param.cpu()
+                        (new_param ** 2).sum().backward()
+                    else:
+                        new_param = new_param.cuda()
```
Comment (albanD) [inline test/test_autograd.py]: nit: for consistency with the use of device below, you can use `.to(device)`

File: test/test_autograd.py [hunk 2]
```diff
@@ -5721,6 +5776,49 @@ def test_copy_(self, device):
...
+                        (new_param ** 2).sum().backward()
+                return grad_output
+
+        # Reentrant starts on GPU thread, finishs on GPU thread
+        x = torch.randn(2, 2, device=device, requires_grad=True)
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # Reentrant starts on CPU thread, finishs on GPU thread
```
Comment (albanD) [inline test/test_autograd.py]: This one will actually finish on the CPU no?
Because the AccumulateGrad for the leaf you create here `new_param = ctx.output_var.detach().requires_grad_()` will run on the CPU.

File: test/test_autograd.py [hunk 3]
```diff
@@ -5721,6 +5776,49 @@ def test_copy_(self, device):
...
+        x = torch.randn(2, 2, requires_grad=True)
+        # set ReentrantFunc node to GPU to emit tasks to GPU queue
+        ReentrantFunc._cpu_mode = False
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # # Reentrant starts on GPU thread, finishs on CPU thread
```
Comment (albanD) [inline test/test_autograd.py]: nit: extra `#`

Also with the same argument as above, this one finished on GPU no?

File: test/test_autograd.py [hunk 4]
```diff
@@ -3568,6 +3575,54 @@ def backward(ctx, input):
...
+        param = torch.rand(10, requires_grad=True)
+
+        def add_gradient_penalty_to_grad(grad):
+            handle.remove()
+            old_param_grad = grad
+            param.grad = None
+            # Add some sort of gradient penalty by directly updating the gradients
+            with torch.enable_grad():
+                g = grad.detach().requires_grad_()
+                new_param = param.detach().requires_grad_()
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/test_autograd.py]: This is needed, otherwise we will ended up with `param`'s AccumulateGrad that is sharing between two GraphTask(one in parent, one in hook), so when you run the AccumulateGrad hook for leaf variable `param`, we are using `param` as the leaf variable for backward call in the hook as well. That will lead to the backward in the hook also tries to run the same `param`'s AccumulateGrad node, which will break thread safety combined with the new Autograd PR.

File: test/test_autograd.py (hunk 2 above)
Comment (wanchaol) [reply test/test_autograd.py]: updated with new tensors instead.

File: test/test_autograd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_autograd.py]: But that should work no?
In particular for things that were discussed in the distributed meeting that were considering moving optimizers into hooks, then trying to add any penalty to the original weights will just deadlock?

File: test/test_autograd.py [hunk 5]
```diff
@@ -5721,6 +5775,46 @@ def test_copy_(self, device):
             z = x.to(torch.bfloat16)
             self.assertTrue(z.requires_grad)
 
+    @onlyCUDA
+    def test_simple_reentrant_cross_device(self, device):
+        class ReentrantFunc(Function):
+            _cpu_mode = True
+            @staticmethod
+            def forward(ctx, x):
+                with torch.enable_grad():
```
Comment (albanD) [inline test/test_autograd.py]: You can also remove the enable_grad and detach() below here.

File: test/test_autograd.py [hunk 6]
```diff
@@ -5721,6 +5775,46 @@ def test_copy_(self, device):
...
+                    ctx.output_var = x * (x + 2) 
+                return ctx.output_var.detach()
```
Comment (pritamdamania87) [inline test/test_autograd.py]: nit: Do we need ctx here? Can we just do `return x * (x + 2)`?

File: test/test_autograd.py [hunk 7]
```diff
@@ -5721,6 +5775,46 @@ def test_copy_(self, device):
...
+                        new_param = torch.randn(2, 2, device=device, requires_grad=True)
+                        (new_param ** 2).sum().backward()
+                return grad_output
+
+        # Reentrant starts on GPU thread, finishs on GPU thread
+        x = torch.randn(2, 2, device=device, requires_grad=True)
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # Reentrant starts on CPU thread, finishs on GPU thread
```
Comment (pritamdamania87) [inline test/test_autograd.py]: Doesn't the backward technically finish on the CPU thread since the last thing we execute is AccumulateGrad for `x`?

File: test/test_autograd.py [hunk 8]
```diff
@@ -5721,6 +5775,46 @@ def test_copy_(self, device):
...
+        x = torch.randn(2, 2, requires_grad=True)
+        # set ReentrantFunc node to GPU to emit tasks to GPU queue
+        ReentrantFunc._cpu_mode = False
+        out = ReentrantFunc.apply(x)
+        out.sum().backward()
+
+        # Reentrant starts on GPU thread, finishs on CPU thread
```
Comment (pritamdamania87) [inline test/test_autograd.py]: This technically finishes on the GPU thread since AccumulateGrad for `x` is executed on the GPU thread?

File: test/test_autograd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/test_autograd.py]: So after offline discussions, this is fine.

File: test/test_autograd.py (hunk 7 above)
Comment (wanchaol) [reply test/test_autograd.py]: Yeah the root backward finishes on CPU as we execute `AccumulateGrad` for `x`, but I think here we want to test the `reentrant` backward calls starts on CPU and finish on GPU? For reentrant backward, it starts on CPU as the root backward on CPU, it then finishes it's backward on GPU since we run `AccumulateGrad` for `new_param` on GPU?

File: test/test_autograd.py (hunk 8 above)
Comment (wanchaol) [reply test/test_autograd.py]: similar comment as above.

**Review by albanD — APPROVED** [review]
Looks good !
